notes.md

```diff
@@ -0,0 +1,179 @@
+- We want the HTTP server to rotate client connections to load balance across multiple instances.
+- We don't want to allow HTTP pipelining.
+- We want HTTP logging (and metrics?).
+  - Access log with URI path, HTTP method, HTTP headers, payload size in bytes, response time, response code, and response headers.
+  - Metrics on how often the TCP connection is closed by the server and the client. How often client drops connections? 
+- We don't need SSL since this service will be behind a Load Balancer that will terminate SSL connections.
+  
+WebSocket (at a minimum): 
+- We want the HTTP connection to transition to an arbitrary WebSocket connection, i.e. there's not a single WS controller with a single path for the whole server.
+- We want to receive websocket frames.
+  - We want to backpressure the sender. For context, say our service proxies frames to other tcp servers. Furthermore,
+    one of those servers is slow. We want the client to slow down their rate without an explicit higher layer protocol
+    message. However, we can optimize this so that a single slow downstream does backpressure all other frames going to
+    other destinations. We maintain a queue of frames and only backpressure when queue is full. (Without an explicit
+    higher layer flow control or protocol other TCP, this will only mitigate the problem. Frames for different
+    downstreams are multiplexed over a single TCP connection, so once queue is full because a single slow consumer than 
+    we block all other frame processing, a.k.a Head-of-line (HOL) blocking. This is partly a fundamental limitation of
+    TCP, i.e. there's no virtual streams. Such limitations are overcome in QUIC and HTTP/3.)
+    Either of these 2 approaches may work:
+    - Only read socket when we finish processing a message. Pub-Sub app may queue up the frame for later processing, or it may immedieiately process it.
+    - Give FS2-Netty a queue to fill and only backpressure when queue is full (or above a high watermark, start reading when below a low watermark).
+- We want to send ws frames. 
+  - When sending any frame, we want to know if it was written to the connection. 
+  - When sending a close frame, then the frame itself should be sent, then the connection should be closed.
+  - We want to know if we're writing too fast and respond to backpressure (namely TCP backpressure).
+- We want to know why a websocket connection closed, what was the close code if any.
+
+WebSocket (extras that will save us time dev time if framework can do these):
+- We want to group connections, label an individual connection, write to a single connection, broadcast to group, close a single connection, and close all connections.
+- We want to know if connection is still alive.
+  - Close connection if it's unresponsive.
+- We want metrics for WS server
+  - Count of different frame types
+  - The size of data frames
+  - Count of different close codes
```

Comment:
I would definitely do this on top of the streams that we produce from fs2-netty. Basically all of those semantics work really nicely and composably downstream because of how Fs2 works.